### PR TITLE
Feature 3206: File download via passthrough

### PR DESCRIPTION
**Pull request by claussni**

### CODE REVIEW

File: Classes/Controller/GetFileController.php [hunk 1]
```diff
@@ -15,31 +15,34 @@
...
  *
- * 3. METS from Goobi.Publication (this extension)
- *   http://localhost/api/3/preview
+ * 3. METS from Kitodo.Publication (this extension)
+ *   http://localhost/api/3/preview/
  *
  *
  * @author    Alexander Bigga <alexander.bigga@slub-dresden.de>
+ * @author    Ralf Claussnitzer <ralf.claussnitzer@slub-dresden.de>
  */
```
Comment (henning-gerhardt) [inline Classes/Controller/GetFileController.php]: In Kitodo.Production we are removing email addresses from author tags as this information will never be updated if the author left the institution.

File: Classes/Controller/GetFileController.php [hunk 2]
```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
...
+                    return $metsXml;
+
+                } else {
+
+                    http_response_code(404);
 
-                return $metsXml;
+                    print('No such document');
+
+                    exit;
```
Comment (henning-gerhardt) [inline Classes/Controller/GetFileController.php]: Is `exit` really this what you want? I think a `return` statement is a better solution.

File: Classes/Controller/GetFileController.php [hunk 3]
```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
 
                     header('Content-Type: text/xml; charset=UTF-8');
 
-                }
+                    return $metsXml;
+
+                } else {
+
+                    http_response_code(404);
```
Comment (henning-gerhardt) [inline Classes/Controller/GetFileController.php]: Is there no solution to set response header with a custom message? I think this should be possible inside TYPO3.

File: Classes/Controller/GetFileController.php [hunk 4]
```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
...
+            if (FALSE !== stripos($value, 'Content-Length')) {
                 header($value);
                 continue;
             }
         }
 
         if ($stream = fopen($path, 'r')) {
 
+            // close active session if any
+            session_write_close();
```
Comment (henning-gerhardt) [inline Classes/Controller/GetFileController.php]: Is session handling not done by TYPO3? So far I see there is no `session_start` inside this code or any other session handling.

File: Classes/Controller/GetFileController.php (hunk 1 above)
Comment (claussni) [inline Classes/Controller/GetFileController.php]: That is actually a good idea. I will see who is in charge for deciding this for Kitodo.Publication.

File: Classes/Controller/GetFileController.php (hunk 4 above)
Comment (chrode) [inline Classes/Controller/GetFileController.php]: Here we are in streaming context. Sessions are kept open. A streaming might take a while so this IMHI correct as sessions should be closed before the streaming starts.

File: Classes/Controller/GetFileController.php [hunk 5]
```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
...
+
+            // stop output buffering
+            ob_end_clean();
+
             fpassthru($stream);
 
             fclose($stream);
```
Comment (chrode) [inline Classes/Controller/GetFileController.php]: Her i'd work with `exit` to quit right after the file streaming is done. There is no use-case where anything should follow the stream.

File: Classes/Controller/GetFileController.php [hunk 6]
```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
...
+                }
 
             case 'attachment':
                 $path = rtrim('http://' . $fedoraHost, "/") . '/fedora/objects/' . $piVars['qid'] . '/datastreams/' . $piVars['attachment'] . '/content';
                 break;
+
             default:
-                break;
+                http_response_code(404);
+                print('No such action');
```
Comment (claussni) [inline Classes/Controller/GetFileController.php]: Turns out `return $somestring` sets the response content respectively. Empty return results in HTML rendered TYPO3 error message.

**Pull request merged**